### PR TITLE
[Tensorflow][docs][to #195]tensorflow blade tutorials

### PR DESCRIPTION
**Pull request by Orion34-lanbo**
As described in [issue](https://github.com/alibaba/BladeDISC/issues/195), adding a tutorial for users to optimize a BERT model with tensorflow-blade's TensorRT optimization.

### CODE REVIEW

File: tensorflow_blade/docs/build_from_source.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
+# Build Tensorflow-Blade from Source
+
+This document introduces how to build Tensorflow-Blade from source.
+As for the preparation part, you can refer to **Prerequisite**, **Checkout the Source**, **Launch a development Docker container**
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/build_from_source.md:4]: link to the three chapter mentioned.

File: tensorflow_blade/docs/build_from_source.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                            - build: build tf blade with regard to the configured part
+                            - test: test tf blade framework.
+                            - package: make tf blade python packages.
+  --device {cpu,gpu}    Build target device
+  --tf {1.15,2.4}       Tensorflow version.
+  --skip-trt            If True, tensorrt will be skipped for gpu build
+  --skip-hie            If True, hie will be skipped for internal build
+  --internal            If True, internal objects will be built
+  --verbose             Show more information in each stage
+  --develop             If True, python wheel develop mode will be installed for local development or debug.
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/build_from_source.md:40]: what is: python wheel develop mode will be installed ?

File: tensorflow_blade/docs/build_from_source.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+```
+
+### A typical build flow for GPU device and Tensorflow-gpu==2.4.0
+ - 1. Create an venv for build
+```bash
+# create a simple venv in docker
+python -m virtualenv ${VENV_PATH} && source ${VENV_PATH}/bin/activate
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/build_from_source.md:47]: Oh, we've not `venv.py` already? if so, why not just eliminate this step?

File: tensorflow_blade/docs/build_from_source.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+python -m pip install -r requirement-tf2.4-cu110.txt
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/build_from_source.md:48]: prefer to use `python3`

File: tensorflow_blade/docs/build_from_source.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+If you want to install the wheel package under develop mode for debug, the following command can be applied.
+```bash
+./build.py ${VENV_PATH} -s package --develop
+```
+
+## NOTE
+Currently only tensorflow-gpu==2.4.0 with CUDA 11.0 is supported for Tensorflow-Blade.
+
+Supports for more tensorflow versions and devices will be coming soon.
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/build_from_source.md:87]: is coming soon.

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
+# Tutorial: Optimize BERT Inference with Tensorflow-Blade
+
+In this tutorial, we show how to optimize BERT model for inference with a few lines code to call tensorrt optimization pass from tensorflow-blade.
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:3]: `TensorRT` pls ~

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+**BERT**, short for Bidirectional Encoder Representations from Transformers, is one of the most popular natural language processing (NLP) model in the world. BERT models and its many varieties have been widely used for language modelling tasks. **However, with great model accuracy comes with a great amount of computations.** The latency of BERT model is quite high, since it has both massive GEMM operations and element-wise operations with lots of redundancy memory accesses, making BERT model difficult to deploy for real-time applications.
+In this tutotial, we use a pre-trained "**bert-base-cased**" BERT model from **Huggingface Transformers**.
+
+**TensorRT** is an SDK for deep learning inference powered by Nvidia, which includes a deep learning inference optimizer and runtime that delivers low latency and high throughput for inference applications.
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:8]: Key logic missing: why not use TensorRT directly?

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+TensorflowBlade can do ahead-of-time optimization with **TensorRT** backend, providing a state-of-the-art inference performance on Nvidia's GPGPUs.
+
+The content of this tutorial is as following.
+  - [Prologue: Prepare](#prologue-prepare)
+  - [Optimize with Tensorflow-Blade's TensorRT backend](#optimize-with-tensorflow-blade)
+  - [Epilogue: Run Inference for origin and optimized model](#epilogue-run-inference-for-origin-and-optimized-model)
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:15]: lol,  `Prologue`/`Epilogue` are really fancy.

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+## Prologue: Prepare
+### environment setup
+These packages are required:
+- tensorflow == 2.4.0
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:20]: tensorflow-gpu == 2.4.0 ??

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+for i in range(len(output_origin)):
+    assert(np.allclose(output_origin[i], output_opt[i], rtol=1e-6, atol=1e-3))
+```
+
+And we can get the following result:
+```shell
+Latency of origin model: 12.40ms
+Latency of optimized model: 3.39ms
+```
+There we can see, using Tensorrt's fp32 optimization, we get a **3.66X** speedup for BERT model inference with acceptable numerical errors.
```
Comment (qiuxiafei) [inline tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:108]: TensorRT not Tensorrt ... pls check all occurrences.

File: tensorflow_blade/examples/Inference/CUDA/BERT/README.md [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+# Run:
+python3 bert_inference_opt.py
+```
+
+## Performance results.
+
+We evaluate origin model and optimized model on T4 GPU, with 1000 inferences.
+The overall execution time of the 1000 inferences is as following.
+
+| TensorFlow  |    XLA     |
```
Comment (qiuxiafei) [inline tensorflow_blade/examples/Inference/CUDA/BERT/README.md:19]: why XLA is here ?

**Review by qiuxiafei — CHANGES_REQUESTED** [review]
Google's [Technical Writing Courses](https://developers.google.com/tech-writing/overview) is a good start for technical writing, if you're interested.

File: tensorflow_blade/README.md [hunk 12]
```diff
@@ -1 +1,62 @@
-# Tensorflow-Blade(WIP)
+# Tensorflow-Blade
+Tensorflow-Blade is a optimization toolkit for tensorflow model inference on multiple devices, like GPGPUs and CPUs.
```
[the remark this answers is not in the document]
Comment (linearhit) [reply tensorflow_blade/README.md:2]: I think such introduction of Tensorflow-Blade is also confusing to the readers (for its relationship with DISC). If we are sure to open source the selection logic of disc/trt, is it possible to define blade as an inference optimization engine that wraps both disc and trt?

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md (hunk 7 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:8]: I think maybe we should explains this in the design of tf2trt part, and reference here will be ok.

File: tensorflow_blade/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md]: done

File: tensorflow_blade/README.md [hunk 13]
```diff
@@ -1 +1,62 @@
...
+## build and use Tensorflow-Blade
+[build and install Tensorflow-Blade](docs/build_from_source.md)
+
+[optimize BERT model with Tensorflow-Blade](docs/tutorials/tensorflow_blade_bert_inference.md)
+
+## design of tf2trt optimization(WIP)
+
+## Make contribution for Tensorflow-Blade
+First ref to [How to make contribution](/docs/contribution.md) to setup the basic develop environment and fork code to your own branch.
+Then we will get to now the code for Tensorflow-Blade.
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md]: typo，should be get to know

File: tensorflow_blade/README.md [hunk 14]
```diff
@@ -1 +1,62 @@
...
+
+### Code tree for Tensorflow-Blade
+All the codes for Tensorflow-Blade are under the **tensorflor\_blade** dir.
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md]: done

File: tensorflow_blade/README.md [hunk 15]
```diff
@@ -1 +1,62 @@
...
+├── tests
+├── tf_blade
+├── venv
+├── version.py
+├── WORKSPACE
+├── workspace0.bzl
+├── workspace1.bzl
+└── workspace2.bzl
+```
+All the files with name containing workspace and build are all used to **bazel build**.
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md]: done

File: tensorflow_blade/README.md [hunk 16]
```diff
@@ -1 +1,62 @@
...
+All the cpp source codes are under **src** dir.
+All the python source codes are under **tf\_blade** dir.
+All the tests files are under **tests** dir.
+
+### CI pipeline for Tensorflow-Blade
+When running CI actions for Tensorflow-Blade, all the configure/build/package as in [build and install Tensorflow-Blade](docs/build_from_source.md) are executed. Also we will run the python lint check for python code and run all the tests under tests dir.
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md]: done

File: tensorflow_blade/README.md [hunk 17]
```diff
@@ -1 +1,62 @@
...
+./build.py ${VENV_PATH} -s check
+```
+The checks will be done using python's flake8/black/mypy modules.
+Please make sure all the python functions or classes have **type hints** as we have done in current code.
+
+As for cpp code, the check will be done by the **pre-commit** tool using clang-format.
+
+ - run unit tests
+```bash
+./build.py ${VENV_PATH} -s test
```
[the remark this answers is not in the document]
Comment (Orion34-lanbo) [reply tensorflow_blade/README.md:64]: `venv` is removed from build.py

File: tensorflow_blade/docs/build_from_source.md (hunk 1 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/build_from_source.md]: done

File: tensorflow_blade/docs/build_from_source.md (hunk 3 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/build_from_source.md]: removed

File: tensorflow_blade/docs/build_from_source.md (hunk 4 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/build_from_source.md]: done

File: tensorflow_blade/docs/build_from_source.md (hunk 5 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/build_from_source.md]: done

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md (hunk 9 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md]: done

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md (hunk 10 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md]: done

File: tensorflow_blade/examples/Inference/CUDA/BERT/README.md (hunk 11 above)
Comment (Orion34-lanbo) [reply tensorflow_blade/examples/Inference/CUDA/BERT/README.md]: typo, fixed

File: tensorflow_blade/README.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (qiuxiafei) [reply tensorflow_blade/README.md:64]: So the `${VENV_PATH}` argument should be removed, too.

File: tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md (hunk 7 above)
Comment (qiuxiafei) [reply tensorflow_blade/docs/tutorials/tensorflow_blade_bert_inference.md:8]: At least it should be explained here why `trtexec` is unable to optimize this model? Or `trtexec` is ok ? -_-!

File: tensorflow_blade/examples/Inference/CUDA/BERT/README.md (hunk 11 above)
Comment (qiuxiafei) [reply tensorflow_blade/examples/Inference/CUDA/BERT/README.md]: lol

**Pull request merged**